File: Algorithms/p121_Best_Time_to_Buy_and_Sell_Stock_1.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        l = len(prices)
        if l<=1:
            return 0
        
        mins = [prices[0]]*(l-1)
        maxs = [prices[l-1]]*(l-1)
        for i in range(1,l-1):
            mins[i] = min(prices[i],mins[i-1])
        for i in range(l-3,-1,-1):
            maxs[i] = max(prices[i+1],maxs[i+1])
        res = 0
        for i in range(0,l-1):
            res = max(maxs[i]-mins[i],res)
        return res

File: Algorithms/p121_Best_Time_to_Buy_and_Sell_Stock/p121_Best_Time_to_Buy_and_Sell_Stock_1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int len = prices.size();
        if(len<=1){
            return 0;
        }
        
        vector<int> mins(len-1,prices[0]);
        vector<int> maxs(len-1,prices[len-1]);
        for(int i=1;i<len-1;++i){
            mins[i] = min(prices[i],mins[i-1]);
        }
        for(int i=len-3;i>=0;--i){
            maxs[i] = max(prices[i+1],maxs[i+1]);
        }
        int res = 0;
        for(int i=0;i<len-1;++i){
            res = max(maxs[i]-mins[i],res);
        }
        return res;
    }
};
